fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/schedule/LoanScheduleComponent.java

```diff
@@ -0,0 +1,136 @@
+                        scheduledLoanInstallment.periodNumber());
+                if (existingInstallment == null) {
+                    final LoanRepaymentScheduleInstallment installment = new LoanRepaymentScheduleInstallment(loan,
+                            scheduledLoanInstallment.periodNumber(), scheduledLoanInstallment.periodFromDate(),
+                            scheduledLoanInstallment.periodDueDate(), scheduledLoanInstallment.principalDue(),
+                            scheduledLoanInstallment.interestDue(), scheduledLoanInstallment.feeChargesDue(),
+                            scheduledLoanInstallment.penaltyChargesDue(), scheduledLoanInstallment.isRecalculatedInterestComponent(),
+                            scheduledLoanInstallment.getLoanCompoundingDetails(), scheduledLoanInstallment.rescheduleInterestPortion(),
+                            scheduledLoanInstallment.isDownPaymentPeriod());
+                    loan.addLoanRepaymentScheduleInstallment(installment);
+                } else {
+                    existingInstallment.setFromDate(scheduledLoanInstallment.periodFromDate());
+                    existingInstallment.setDueDate(scheduledLoanInstallment.periodDueDate());
+                    existingInstallment.setPrincipal(scheduledLoanInstallment.principalDue());
+                    existingInstallment.setInterestCharged(scheduledLoanInstallment.interestDue());
+                    existingInstallment.setFeeChargesCharged(scheduledLoanInstallment.feeChargesDue());
+                    existingInstallment.setPenaltyCharges(scheduledLoanInstallment.penaltyChargesDue());
+
+                    existingInstallment.setRecalculatedInterestComponent(scheduledLoanInstallment.isRecalculatedInterestComponent());
+                    existingInstallment.setLoanCompoundingDetails(scheduledLoanInstallment.getLoanCompoundingDetails());
+                    existingInstallment.setRescheduleInterestPortion(scheduledLoanInstallment.rescheduleInterestPortion());
+                    existingInstallment.setDownPayment(scheduledLoanInstallment.isDownPaymentPeriod());
+                }
+            }
+        }
+        // Review Installments removed
+        List<LoanRepaymentScheduleInstallment> removeInstallments = new ArrayList<>();
+        for (final LoanRepaymentScheduleInstallment existingInstallment : existingInstallments) {
+            if (findByInstallmentNumber(periods, existingInstallment.getInstallmentNumber()) == null) {
+                removeInstallments.add(existingInstallment);
+            }
+        }
+        if (!removeInstallments.isEmpty()) {
+            existingInstallments.removeAll(removeInstallments);
+        }
+
+        loan.updateLoanScheduleDependentDerivedFields();
+        loan.updateLoanSummaryDerivedFields();
+    }
+
+    public void updateLoanSchedule(Loan loan, final Collection<LoanRepaymentScheduleInstallment> installments) {
```

Comment:
We can modify the param type from Collection to List... our installments are always a list